plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/HudiSplitSource.java

```diff
@@ -31,56 +37,54 @@
 import org.apache.hudi.common.table.HoodieTableMetaClient;
-import org.apache.hudi.common.util.HoodieTimer;
 
-import java.util.ArrayDeque;
-import java.util.ArrayList;
-import java.util.Deque;
 import java.util.List;
 import java.util.Map;
 import java.util.concurrent.CompletableFuture;
-import java.util.concurrent.Executors;
-import java.util.concurrent.ScheduledExecutorService;
-import java.util.concurrent.ScheduledFuture;
-import java.util.concurrent.TimeUnit;
+import java.util.concurrent.ExecutorService;
 
+import static com.google.common.util.concurrent.Futures.immediateFailedFuture;
+import static com.google.common.util.concurrent.MoreExecutors.directExecutor;
+import static io.airlift.concurrent.MoreFutures.toCompletableFuture;
+import static io.trino.plugin.hudi.HudiSessionProperties.getMinimumAssignedSplitWeight;
+import static io.trino.plugin.hudi.HudiSessionProperties.getStandardSplitWeightSize;
 import static io.trino.plugin.hudi.HudiSessionProperties.isHudiMetadataEnabled;
+import static io.trino.plugin.hudi.HudiSessionProperties.isSizeBasedSplitWeightsEnabled;
 import static io.trino.plugin.hudi.HudiSessionProperties.shouldSkipMetaStoreForPartition;
-import static io.trino.plugin.hudi.HudiUtil.getMetaClient;
-import static java.lang.String.format;
-import static java.lang.Thread.currentThread;
-import static java.lang.Thread.sleep;
-import static java.util.concurrent.CompletableFuture.completedFuture;
+import static io.trino.spi.StandardErrorCode.GENERIC_INTERNAL_ERROR;
+import static java.util.concurrent.CompletableFuture.runAsync;
 import static java.util.stream.Collectors.toList;
 
 public class HudiSplitSource
         implements ConnectorSplitSource
 {
-    private static final Logger log = Logger.get(HudiSplitSource.class);
-    private static final long IDLE_WAIT_TIME_MS = 10;
-    private final HudiFileListing hudiFileListing;
-    private final Deque<ConnectorSplit> connectorSplitQueue;
-    private final ScheduledFuture splitLoaderFuture;
+    private final AsyncQueue<ConnectorSplit> queue;
+
+    private TrinoException trinoException;
 
     public HudiSplitSource(
             ConnectorSession session,
             HiveMetastore metastore,
             Table table,
             HudiTableHandle tableHandle,
             Configuration conf,
-            Map<String, HiveColumnHandle> partitionColumnHandleMap)
+            Map<String, HiveColumnHandle> partitionColumnHandleMap,
+            ExecutorService executor,
+            int maxSplitsPerSecond,
+            int maxOutstandingSplits)
     {
         boolean metadataEnabled = isHudiMetadataEnabled(session);
         boolean shouldSkipMetastoreForPartition = shouldSkipMetaStoreForPartition(session);
-        HoodieTableMetaClient metaClient = tableHandle.getMetaClient().orElseGet(() -> getMetaClient(conf, tableHandle.getBasePath()));
+        HoodieTableMetaClient metaClient = buildTableMetaClient(conf, tableHandle.getBasePath());
```

Comment:
Won't this instantiate meta client per split? Since the metaclient remains the same for the table, so we don't really need to instantiate per split right?
May I know why getMetaClient() was removed from table handle?

Reply:
yeah makes sense.